[Security] Ability to add roles in `form_login_ldap` by ldap group

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #51225
| License       | MIT

This PR adds a way for setting roles in `form_login_ldap` based on LDAP configuration.
~Please note that it is based on SF6.4, but may be changed to 7.1 if already in feature freeze period.~ => Rebased for targeting SF 7.1

* [x] Feature
* [x] Tests
* [ ] Documentation

OskarStark commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/UserProvider/LdapFactory.php:
```diff
@@ -40,6 +41,7 @@ public function create(ContainerBuilder $container, string $id, array $config)
             ->replaceArgument(6, $config['filter'])
             ->replaceArgument(7, $config['password_attribute'])
             ->replaceArgument(8, $config['extra_fields'])
+            ->replaceArgument(9, $roleFetcher)
```
```suggestion
            ->replaceArgument(9, $$config['role_fetcher'] ? new Reference($config['role_fetcher']) : null)
```

I think we can keep it in one line

Spomky replied:
Indeed. Comment addressed.

OskarStark commented:
> Please note that it is based on SF6.4, but may be changed to 7.1 if already in feature freeze period.

Yes it will be for 7.1, but you can already rebase on and target 7.0 if you like

nicolas-grekas commented on src/Symfony/Component/Ldap/Security/LdapUserProvider.php:
```diff
@@ -154,7 +156,13 @@ protected function loadUser(string $identifier, Entry $entry): UserInterface
         }
 
-        return new LdapUser($entry, $identifier, $password, $this->defaultRoles, $extraFields);
+        $roles = $this->defaultRoles;
+
+        if (null !== $this->roleFetcher) {
+            $roles = $this->roleFetcher->fetchRoles($entry);
+        }
```
```suggestion
        $roles = $this?->roleFetcher->fetchRoles($entry) ?? $this->defaultRoles;
```

nicolas-grekas reviewed (commented):
What would be the typical implementation for a role fetcher? Can we provide one that e.g. reads some conventional entries from the `Entry`?

OskarStark commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/JsonLoginLdapTest.php:
```diff
@@ -22,4 +29,42 @@ public function testKernelBoot()
+                return $this->getArrayCopy();
+            }
+        };
+        $queryMock = $this->getMockBuilder(QueryInterface::class)->getMock();
```
```suggestion
        $queryMock = $this->createMock(QueryInterface::class);
```

Same for the other

Spomky replied:
Many thanks. COmment addressed.

Spomky commented:
> What would be the typical implementation for a role fetcher? Can we provide one that e.g. reads some conventional entries from the `Entry`?

Hello @nicolas-grekas,
I added a the classes [AssignDefaultRoles](https://github.com/symfony/symfony/pull/52181/files#diff-6a35d6457b8fa6e2ea7fa4b90285289999e4b0922d0b7c1e88b3f4ea223888d0) and [MemberOfRoles](https://github.com/symfony/symfony/pull/52181/files#diff-e1ad6f8ddf99db72e72a19f588506a70d020e2e19e9bb28b7375705a28420b33). The latter uses the `ismemberof` attribute that seems to be widely used. Note that there is no specification for such group attribute; the [RFC4519](https://datatracker.ietf.org/doc/html/rfc4519) only describes static groups.

OskarStark commented on src/Symfony/Component/Ldap/Security/AssignDefaultRoles.php:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * @param string[] $roles
+     */
+    public function __construct(private array $roles)
```
```suggestion
    public function __construct(
    private array $roles)
```

newline please

Spomky replied:
Hi,
I modified this file that now have arguments on a dedicated line.

MDeveloping commented:
Did I understand this correct, that there will be an easy way, to read the memberOf of a LDAP User if logging in and than set a role?
Did I also understand it correct, that will be available at Symphony 7.2 ad that a documentation is missing for now on Symfony Docs?

welcoMattic commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/JsonLoginLdap/config.yml:
```diff
@@ -21,7 +22,7 @@ security:
                 default_roles: ROLE_USER
                 uid_key: uid
-                extra_fields: ['email']
+                #extra_fields: ['email']
```
Still needed or should be removed?

Spomky replied:
I am not sure. I will check if still needed.

welcoMattic commented on src/Symfony/Component/Ldap/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+7.3
+---
+
+ * Add ability to fetch roles
```
```suggestion
 * Add `RoleFetcherInterface` to allow roles fetching at user loading
```

@Spomky it's a suggestion, you can rephrase it if needed. The goal is to make it more precise.

welcoMattic reviewed (approved):
2 minor comments, otherwise it's 👍 for me

fabpot commented on src/Symfony/Component/Ldap/Security/LdapUserProvider.php:
```diff
@@ -147,7 +150,9 @@ protected function loadUser(string $identifier, Entry $entry): UserInterface
             $extraFields[$field] = $this->getAttributeValue($entry, $field);
         }
 
-        return new LdapUser($entry, $identifier, $password, $this->defaultRoles, $extraFields);
```
IIUC, `$this->defaultRoles` is not used when setting a role fetcher. So, it's either one or the other.

Instead of adding a new argument to the constructor, maybe we can make `$defaultRoles` and array or a `RoleFetcherInterface` instance?

Spomky replied:
You are right. I will change it.

Spomky replied:
`$defaultRoles` is now an array or an instance.
The bundle configuration is updated accordingly (including the test).

fabpot commented on src/Symfony/Bundle/SecurityBundle/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Add `Security::isGrantedForUser()` to test user authorization without relying on the session. For example, users not currently logged in, or while processing a message from a message queue
  * Add encryption support to `OidcTokenHandler` (JWE)
+ * Add ability to fetch Ldap roles
```
```suggestion
 * Add ability to fetch LDAP roles
```

Spomky committed:
[Security] Ability to add roles in form_login_ldap by ldap group

This update allows LDAP to fetch roles for a given user entry by using the new RoleFetcherInterface. The LdapUserProvider class has been adjusted to use this new functionality.

fabpot commented:
Thank you @Spomky.
